### DISCUSSION TITLE
Is Grott a proxy server?

### DISCUSSION THREAD

**Discussion by andrewfg** [johanmeijer/grott#351] [Q&A] [answered]
Can someone please clarify the following for me..

> The purpose of Grott is to read, parse and forward the raw metrics as they are sent to Growatt servers

If I understand correctly, Grott interposes itself between the inverter and the Growatt server. Does this mean that it is acting as a proxy server on the LAN in some way? Or is it interposing itself in some other manner?

**Comment by johanmeijer** [comment, marked as answer]
Indeed it is a proxy server. You can find more info in the wiki.

**Comment by rany2** [comment]
If you're struggling to understand it, consider this:

```
Your Inverter <---> Grott (Reads data sent by inverter and passes it to Growatt) <---> Growatt's Servers
```

So in essence the inverter uses Grott as an intermediary to connect to the Growatt servers and by doing so Grott gets to read the messages between the inverter and Growatt. Hope it helps. (Edit: should be noted there is an exception only in GrottServer mode where Grott assumes all the roles of Growatt's server but the principle/mode of operation is the same)

**Comment by andrewfg** [comment]
Thank you both for the information.

I am a Java developer who uses openHAB and who has written several binding integrations into openHAB. I shall probably be buying a Growatt inverter next week. And I shall be seeking to integrate it into OH via Grott.

I see that Grott can forward data via MQTT to (say) the MQTT server running on OH. In that configuration one would probably have a separate RPi running Grott, and another RPi running OH. It probably requires a lot of configuration work mapping the Grott data via MQTT into OH format. Which is probably not very user friendly for the average Joe.

So I think that I shall attempt to make a tighter integration between Grott and OH; for example by running Grott and OH on the same RPi  and passing data within the two applications on the same computer. I will have a look into your code to see what can be done. But if you have any thoughts I would much appreciate it. :)

**Comment by andrewfg** [comment]
^
So in a nutshell here is roughly what I would do. Basically I would use Grott with your existing HTTP client extension application to POST JSON data to a server url on the OpenHAB main UI web server that is running on the same PC e.g. http://127.0.0.1:8080/grott. And I would write a new binding for openHab that would do the following..
- The binding implements a Bridge thing which receives the data from Grott
- Bridge checks if Grott is installed and informs user via OH binding UI with an error if not
- Bridge configuration UI reads and possibly writes various relevant Grott configuration file entries
- Bridge implements a switch channel that can start Grott (and maybe stop it) and show its actual state
- Bridge creates and opens an HTTP server applet to receive Grott JSON data POSTs
- Bridge receives and parses JSON and dynamically creates an 'Inverter' thing for each such inverter found
- Each 'Inverter' thing dynamically creates a state channel for each of the attributes in the respective inverter's JSON data, and sets it state accordingly.

EDIT: I would need to make sure that the Grott proxy server is insulated from the OpenHAB main UI server running on the same machine.

**Comment by johanmeijer** [comment]
Integration with openhab is a good idea. Have you seen that Grott supports extensions (https://github.com/johanmeijer/grott/wiki/Extensions).

 It it very easy to create an extension which does a http request which can send json data to a openhab http listen. There is a extension example with http request in the samples (library.https://github.com/johanmeijer/grott/blob/Beta-(2.8.x)/examples/Extensions/grottext.py)

At this moment there are also extension for Home Assistent and generating csv (excel) files.

**Comment by andrewfg** [comment]
> There is a extension example with http request 

Yes, I think that I can use that one. Many thanks.

Apropos the JSON payload, I saw your examples of typical JSON contents, with JSON elements for current/voltage/power/energy for both PV strings and for the grid. Since my solar system will also have a battery, do you know if there are any JSON elements for the battery data?

**Comment by johanmeijer** [comment]
What type inverter do you have? You probably have to specify invtype= in the grott.ini. you can specify spf, sph, spa, min or tl3. 

I tried to define also most of the battery values in the records. If not everything is there we can try to find it.

**Reply by andrewfg** [reply]
Its an SPH. Many thanks.
